src/ZoneServer/World/Maps/Pathfinder.cs

```diff
@@ -0,0 +1,266 @@
+		/// </summary>
+		/// <param name="distance"></param>
+		/// <param name="entitySize"></param>
+		/// <returns></returns>
+		private int GetGridScale(double distance, SizeType entitySize)
+		{
+			var radius = _entitySizeRadius[entitySize];
+			// radius * 2 is the maximum guaranteed value to not cause entities
+			// to go through objects. The client cannot handle this, so it
+			// displays as if entities are teleporting around. Values higher
+			// may cause this visual issue but it may also be desirable due
+			// to computational reasons.
+			return (int)Math.Min(distance / 2, radius * 2);
+		}
+
+		/// <summary>
+		/// Reconstructs the path from the given position.
+		/// </summary>
+		/// <param name="cameFrom"></param>
+		/// <param name="position"></param>
+		/// <returns></returns>
+		private List<Position> ReconstructPath(Dictionary<Position, Position> cameFrom, Position position)
+		{
+			var totalPath = new List<Position>();
+			totalPath.Add(position);
+
+			while (cameFrom.ContainsKey(position))
+			{
+				var nextNode = cameFrom[position];
+
+				position = nextNode;
+				totalPath.Add(position);
+			}
+
+			totalPath.Reverse();
+			return totalPath;
+		}
+
+		/// <summary>
+		/// Calculates the heuristic cost from position a to position b.
+		/// </summary>
+		/// <param name="a"></param>
+		/// <param name="b"></param>
+		/// <returns></returns>
+		private float Heuristic(Position a, Position b)
+		{
+			// Euclidean Distance
+			var dx = Math.Abs(a.X - b.X);
+			var dy = Math.Abs(a.Y - b.Y);
+			var dz = Math.Abs(a.Z - b.Z);
+			return (float)Math.Sqrt(Math.Pow(dx, 2) + Math.Pow(dy,2) + Math.Pow(dz, 2));
+		}
+
+		/// <summary>
+		/// Gets the neighboring positions for a given position under a given
+		/// grid scale.
+		/// </summary>
+		/// <param name="pos"></param>
+		/// <param name="entitySize"></param>
+		/// <param name="gridScale"></param>
+		/// <returns>A list of neighboring positions.</returns>
+		private List<Position> GetNeighbors(Position pos, SizeType entitySize, int gridScale)
+		{
+			// Scale is too low
+			if (gridScale <= 0)
+				return new List<Position>();
+
+			var radius = _entitySizeRadius[entitySize];
+			var neighbors = new List<Position>();
+			var d = gridScale;
+			var directions = new (int X, int Z)[] { (-d, 0), (d, 0), (0, -d), (0, d), (-d, d), (-d, -d), (d, -d), (d, d) };
+
+			// Search in all adjacent directions
+			foreach (var dir in directions)
+			{
+				var neighbor = new Position(pos.X + dir.X, 0, pos.Z + dir.Z);
+
+				// Position at neighbor is valid
+				if (_ground.TryGetHeightAt(neighbor, out var height))
+				{
+					// Our entity can stand there
+					if (_ground.IsValidCirclePosition(neighbor, radius))
+					{
+						neighbor.Y = height;
+						neighbors.Add(neighbor);
+					}
+				}
+			}
+
+			// No adjacent neighbors found
+			if (neighbors.Count == 0)
+			{
+				// Try again recursively with lower scale
+				return this.GetNeighbors(pos, entitySize, gridScale/2);
```

Comment:
It would be preferential if you used auto formatting on your code, as code that is not formatted based on our conventions will create unnecessary whitespace changes the next time someone auto formats a file. Like here with the missing spaces between `gridScale/2`. It's technically a minor thing of course, but it can make reading PRs or double checking commits a whole lot more annoying when you have to find the actual changes first.